ethereum/contracts/IncentivizedReceiveChannel.sol

```diff
@@ -0,0 +1,173 @@
+            require(
+                commitmentContents.messages[i].payload.length <=
+                    MAX_PAYLOAD_BYTE_SIZE,
+                "message payload bytesize exceeds maximum payload size"
+            );
+        }
+        return true;
+    }
+
+    function processCommitmentContents(
+        CommitmentContents memory commitmentContents
+    ) internal {
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            // Check message nonce is correct and increment nonce for replay protection
+            Message memory message = commitmentContents.messages[i];
+            require(message.nonce == lastProcessedNonce + 1, "invalid nonce");
+
+            lastProcessedNonce = lastProcessedNonce + 1;
+
+            // Deliver the message to the destination
+            // Delivery will have fixed maximum gas allowed for the destination app.
+            address targetApplicationAddress = message.targetApplicationAddress;
+            uint256 allowedGas = MAX_GAS_PER_MESSAGE;
+            bytes memory callInput = message.payload;
+
+            bool result;
+            uint256 callInputLength = callInput.length;
+            assembly {
+                let callInputDataPointer := add(callInput, 32) // Move 32 ahead of callInput length slot to get to actual data slot (TODO: verify this works in all cases)
+
+                // Dispatch call to the receiver - it is expected to be fire and forget. If the call reverts, runs out of gas, error,
+                // etc, its the fault of the sender
+                result := call(
+                    allowedGas, // Allowed gas
+                    targetApplicationAddress, // To addr
+                    0, // No ether value being sent
+                    callInputDataPointer, // Inputs are stored at callInputDataPointer
+                    callInputLength, // Input length
+                    callInputDataPointer, // Store output over input (saves space)
+                    callInputLength // Outputs can be as long as input length
+                )
+            }
+
+            emit MessageDelivered(message.nonce, result);
+        }
+    }
+
+    function validateCommitment(
+        Commitment memory commitment,
+        CommitmentContents memory commitmentContents
+    ) internal returns (bool) {
+        bytes32 commitmentHash;
+        for (uint256 i = 0; i < commitmentContents.messages.length; i++) {
+            if (i == 0) {
+                commitmentHash = hashMessage(commitmentContents.messages[i]);
+            } else {
+                bytes32 messageHash =
+                    hashMessage(commitmentContents.messages[i]);
+                commitmentHash = keccak256(
+                    abi.encodePacked(commitmentHash, messageHash)
+                );
+            }
+        }
+        return
+            keccak256(abi.encodePacked(commitmentHash)) ==
+            keccak256(abi.encodePacked(commitment.commitmentHash));
```

Comment:
Why do we need to hash again here? We could just do the following, right?

```suggestion
            commitmentHash == commitment.commitmentHash;
```

Reply:
ye @denalimarsh is working on fixing this.

Reply:
Addressed in https://github.com/Snowfork/polkadot-ethereum/pull/219/commits/35478272b36387ce85ca211d150cf839f39abb57